pkg/splunk/enterprise/util.go

```diff
@@ -882,3 +866,85 @@ func setupAppInitContainers(client splcommon.ControllerClient, cr splcommon.Meta
 		}
 	}
 }
+
+// SetLastAppInfoCheckTime sets the last check time to current time
+func SetLastAppInfoCheckTime(appInfoStatus *enterprisev1.AppDeploymentContext) {
+	scopedLog := log.WithName("SetLastAppInfoCheckTime")
+	currentEpoch := time.Now().Unix()
+
+	scopedLog.Info("Setting the LastAppInfoCheckTime to current time", "current epoch time", currentEpoch)
+
+	appInfoStatus.LastAppInfoCheckTime = currentEpoch
+}
+
+// HasAppRepoCheckTimerExpired checks if the polling interval has expired
+func HasAppRepoCheckTimerExpired(appFrameworkRef *enterprisev1.AppFrameworkSpec, appInfoContext *enterprisev1.AppDeploymentContext) bool {
+	scopedLog := log.WithName("hasAppRepoCheckTimerExpired")
+	currentEpoch := time.Now().Unix()
+	scopedLog.Info("Checking if the app repo polling interval has expired", "LastAppInfoCheckTime", strconv.FormatInt(appInfoContext.LastAppInfoCheckTime, 10), "current epoch time", strconv.FormatInt(currentEpoch, 10))
+	return appInfoContext.LastAppInfoCheckTime+int64(appFrameworkRef.AppsRepoPollInterval) <= currentEpoch
+}
+
+// GetNextRequeueTime gets the next reconcile requeue time based on the appRepoPollInterval.
+// There can be some time elapsed between when we frst set lastAppInfoCheckTime and when the CR is in Ready state.
+// Hence we need to subtracct the delta time elapsed from the actual polling interval,
+// so that the next reconile would happen at the right time.
+func GetNextRequeueTime(appRepoPollInterval, lastCheckTime int64) time.Duration {
+	scopedLog := log.WithName("GetNextRequeueTime")
+	currentEpoch := time.Now().Unix()
+
+	var nextRequeueTimeInSec int64
+	nextRequeueTimeInSec = appRepoPollInterval - (currentEpoch - lastCheckTime)
+
+	scopedLog.Info("Getting next requeue time", "LastAppInfoCheckTime", lastCheckTime, "Current Epoch time", currentEpoch, "nextRequeueTimeInSec", nextRequeueTimeInSec)
+
+	return time.Second * (time.Duration(nextRequeueTimeInSec))
+}
+
+// initAndCheckAppInfoStatus initializes the S3Clients and checks the status of apps on remote storage.
+func initAndCheckAppInfoStatus(client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkConf *enterprisev1.AppFrameworkSpec, appStatusContext *enterprisev1.AppDeploymentContext) error {
+	scopedLog := log.WithName("initAndCheckAppInfoStatus").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	var err error
+	// Register the S3 clients specific to providers if not done already
+	// This is done to prevent the null pointer dereference in case when
+	// operator crashes and comes back up and the status of app context was updated
+	// to match the spec in the previous run.
+	initAppFrameWorkContext(appFrameworkConf, appStatusContext)
+
+	//check if the apps need to be downloaded from remote storage
+	if HasAppRepoCheckTimerExpired(appFrameworkConf, appStatusContext) || !reflect.DeepEqual(appStatusContext.AppFrameworkConfig, *appFrameworkConf) {
```

Comment:
I think the timer should cover the boundary cases. Ideally, the polling time should be covering from the status context?

Reply:
The check for reflect.DeepEqual takes care of the case where if someone modifies the CR and applies it before the timer expires, then we should check the status immediately. And about the polling time, I think it doesn't matter where you pick it up from. On a personal front, I think its better to check it from the CR spec than from the context since there might be some delay in writing the status to the etcd.

Reply:
My comment was about the boundary conditions, say, somebody configures like 1 second, OR 100 days, in which, we are using default min/max :-)